File: robotic_hw/robotic_hw/load_cell_data_publisher.py
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from functools import partial
from robotic_hw_interfaces.srv import GetLoadCellData
from robotic_hw_interfaces.msg import LoadCellDataArray, LoadCellData

class LoadCellDataPublisher(Node):
    """ Calls GetLoadCellData services to get the latest load cell data and publishes it at specified frequency.
        The services are specified as parameters to this node, allowiing the node to publish information from any number of load cells.
        Calling services and publishing are decoupled in different timers. This means old data may be published several times if the service call is slow.
    """

    def __init__(self) -> None:
        super().__init__('load_cell_data_publisher')


        # Create parameter for service names
        self.declare_parameter('service_names', ['get_load_cell_data']) 
        self.service_names = self.get_parameter('service_names').get_parameter_value().string_array_value

        # Create parameter for publish frequency (default 500 Hz)
        self.declare_parameter('publish_frequency', 500.0)
        self.publish_frequency = self.get_parameter('publish_frequency').get_parameter_value().double_value

        # Create parameter for service call frequency (default 500 Hz)
        self.declare_parameter('service_call_frequency', 200.0)
        self.service_call_frequency = self.get_parameter('service_call_frequency').get_parameter_value().double_value

        # Create a client for each service
        self.load_cell_clients = []
        for service_name in self.service_names:
            client = self.create_client(GetLoadCellData, service_name)
            self.load_cell_clients.append(client)
            while not client.wait_for_service(timeout_sec=1.0):
                self.get_logger().warn(f'{service_name} service not available, waiting...')
            self.get_logger().info(f'{service_name} service available.')
        self.futures = set() # Store the active futures so they are not garbage collected before they complete

        # Create request object (same for all services)
        self.req = GetLoadCellData.Request()

        # Create publisher
        self.publisher_ = self.create_publisher(LoadCellDataArray, 'load_cells', qos_profile_sensor_data)
        
        # Define message to publish. Service call timer will update this message and publish_timer will publish it.
        self.load_cell_data_array_msg = LoadCellDataArray() 
        self.load_cell_data_array_msg.load_cells = [LoadCellData(data_ready=False) for _ in self.service_names]

        # Create timer to call service and store results
        self.service_call_timer = self.create_timer(1/self.service_call_frequency, self.service_call_timer_callback)

        # Create timer to publish data
        self.publish_timer = self.create_timer(1/self.publish_frequency, self.publish_timer_callback)

    def publish_timer_callback(self) -> None:
        """ Publishes the latest load cell data.
        """
        self.publisher_.publish(self.load_cell_data_array_msg)

    def service_call_timer_callback(self) -> None:
        """ Calls the GetLoadCellData service asynchronously for each load cell.
            The future callback will process the result and store it in the LoadCellDataArray message.
        """
        # For each client, call the service asynchronously
        for i, client in enumerate(self.load_cell_clients): 
            if client.service_is_ready():
                future = client.call_async(self.req)
                future.add_done_callback(partial(self.future_callback, client_index=i))
                self.futures.add(future)
            else:
                # If the service is not available, log an error and set the future to None
                # Set data_ready to False for that LoadCellData message
                self.get_logger().error(f'Load cell data service {i} not available.')
                self.load_cell_data_array_msg.load_cells[i] = LoadCellData(data_ready=False) 
                return

    def future_callback(self, future, client_index: int = None) -> None:
        """ When the service call is done, this function is called to process the result.
            We store the result in the LoadCellDataArray message  at index client_index.
        """
        try:
            self.load_cell_data_array_msg.load_cells[client_index] = future.result().load_cell_data
            self.futures.remove(future)
        except Exception as e:
            self.get_logger().error(f'Exception while calling service: {e}')
            return
        
def main(args=None) -> None:
    rclpy.init(args=args)
    load_cell_data_client = LoadCellDataPublisher()

    try:
        rclpy.spin(load_cell_data_client)
    except KeyboardInterrupt:
        pass
    finally:
        rclpy.try_shutdown()
        load_cell_data_client.destroy_node()

    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: robotic_hw/src/load_cell_data_publisher.cpp
#include "load_cell_data_publisher.hpp"
#include <chrono>
#include <functional>

using namespace std::chrono_literals;

LoadCellDataPublisher::LoadCellDataPublisher() : Node("load_cell_data_publisher")
{
    // Create parameter for names of services to call
    this->declare_parameter<std::vector<std::string>>("service_names", {"get_load_cell_data"});
    this->get_parameter("service_names", service_names_);

    // Create parameter for the publish frequency (500 Hz by default)
    this->declare_parameter<double>("publish_frequency", 500.0);
    this->get_parameter("publish_frequency", publish_frequency_);

    // Create parameter for the service call frequency 
    this->declare_parameter<double>("service_call_frequency", 200.0);
    this->get_parameter("service_call_frequency", service_call_frequency_);

    // Create clients for each service
    for (const auto& service_name : service_names_)
    {
        auto client = this->create_client<robotic_hw_interfaces::srv::GetLoadCellData>(service_name);
        load_cell_clients_.push_back(client);
        while (!client->wait_for_service(1s))
        {
            RCLCPP_WARN(this->get_logger(), "%s service not available, waiting...", service_name.c_str());
        }
        RCLCPP_INFO(this->get_logger(), "%s service available.", service_name.c_str());
    }

    load_cell_data_array_msg_.load_cells.resize(service_names_.size());

    publisher_ = this->create_publisher<robotic_hw_interfaces::msg::LoadCellDataArray>("load_cells", 10);
    publish_timer_ = this->create_wall_timer(
        std::chrono::milliseconds(static_cast<int>(1000 / publish_frequency_)),
        std::bind(&LoadCellDataPublisher::publish_timer_callback, this));

    service_call_timer_ = this->create_wall_timer(
        std::chrono::milliseconds(static_cast<int>(1000 / service_call_frequency_)),
        std::bind(&LoadCellDataPublisher::service_call_timer_callback, this));
}

void LoadCellDataPublisher::publish_timer_callback()
{
    publisher_->publish(load_cell_data_array_msg_);
}

void LoadCellDataPublisher::service_call_timer_callback()
{
    for (size_t i = 0; i < load_cell_clients_.size(); ++i)
    {
        if (!load_cell_clients_[i]->service_is_ready())
        {
            RCLCPP_ERROR(this->get_logger(), "Load cell data service %zu not available.", i);
            load_cell_data_array_msg_.load_cells[i].data_ready = false;
            continue;
        }

        auto request = std::make_shared<robotic_hw_interfaces::srv::GetLoadCellData::Request>();
        auto result_future = load_cell_clients_[i]->async_send_request(request,
            [this, i](std::shared_future<std::shared_ptr<robotic_hw_interfaces::srv::GetLoadCellData::Response>> future) {
            this->handle_service_response(future, i); // Use 'this' to call a member function
            });
    }
}

void LoadCellDataPublisher::handle_service_response(rclcpp::Client<robotic_hw_interfaces::srv::GetLoadCellData>::SharedFuture future, size_t index)
{
    auto response = future.get();
    load_cell_data_array_msg_.load_cells[index] = response->load_cell_data;
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LoadCellDataPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
